runtime/drivers/pinot/sqldriver/driver.go

```diff
@@ -0,0 +1,293 @@
+package sqldriver
+
+import (
+	"context"
+	"database/sql"
+	sqlDriver "database/sql/driver"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"math/big"
+	"net/url"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/startreedata/pinot-client-go/pinot"
+)
+
+type pinotDriver struct{}
+
+func (d *pinotDriver) Open(dsn string) (sqlDriver.Conn, error) {
+	address, headers, err := ParseDSN(dsn)
+	if err != nil {
+		return nil, err
+	}
+	pinotConn, err := pinot.NewWithConfig(&pinot.ClientConfig{
+		ExtraHTTPHeader: headers,
+		ControllerConfig: &pinot.ControllerConfig{
+			ControllerAddress: address,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	// We have joins and nested queries which are supported by multistage engine
+	pinotConn.UseMultistageEngine(true)
+	return &conn{pinotConn: pinotConn}, nil
+}
+
+func init() {
+	sql.Register("pinot", &pinotDriver{})
+}
+
+type conn struct {
+	pinotConn *pinot.Connection
+}
+
+func (c *conn) Prepare(query string) (sqlDriver.Stmt, error) {
+	return nil, fmt.Errorf("unsupported")
+}
+
+func (c *conn) Close() error {
+	return nil
+}
+
+func (c *conn) Begin() (sqlDriver.Tx, error) {
+	return nil, fmt.Errorf("unsupported")
+}
+
+func (c *conn) QueryContext(ctx context.Context, query string, args []sqlDriver.NamedValue) (sqlDriver.Rows, error) {
+	if len(args) > 0 {
+		q, err := completeQuery(query, args)
+		if err != nil {
+			return nil, err
+		}
+		query = q
+	}
+	// TODO: cancel the query if ctx is done
```

Comment:
Still seems like an important TODO to solve since we do so many query cancellations